tools/cpp/modules_tools/common/common.cc

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.h"
+#include "json.hpp"
+
+#include <iostream>
+
+ModuleDep parse_ddi(std::istream &ddi_stream) {
+  ModuleDep dep{};
+  std::string ddi_string((std::istreambuf_iterator<char>(ddi_stream)),
+                         std::istreambuf_iterator<char>());
+  JsonValue data = parse_json(ddi_string);
+  if (!data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+
+  auto rules_data = data.as_object().at("rules");
+  if (!rules_data.is_array()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto rules = rules_data.as_array();
+  if (rules.size() > 1) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  if (rules.empty()) {
+    return dep;
+  }
+  auto rule_data = rules[0];
+  if (!rule_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto rule = rule_data.as_object();
+  auto provides_data = rule["provides"];
+  if (!provides_data.is_array()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto provides = provides_data.as_array();
+  if (provides.size() > 1) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  if (provides.size() == 1) {
```

Comment:
This is fine for C++, but Fortran can provide more than one module per TU. (Probably fine for Bazel today, but maybe worth a comment.)